python/python/lance/util.py

```diff
@@ -40,3 +45,71 @@ def sanitize_ts(ts: ts_types) -> datetime:
     elif not isinstance(ts, datetime):
         raise TypeError(f"Unrecognized version timestamp {ts} of type {type(ts)}")
     return ts
+
+
+class KMeans:
+    def __init__(self, k: int, metric_type: str = "l2", max_iters: int = 50):
+        """Create a KMeans model.
+
+        Parameters
+        ----------
+        k: int
+            The number of clusters to create.
+        metric_type: str, default="l2"
+            The metric to use for calculating distances between vectors.
+            Supported distance metrics: "l2", "cosine", "dot"
+        max_iters: int
+            The maximum number of iterations to run the KMeans algorithm. Default: 50.
+        """
+        self.k = k
+        self._metric_type = metric_type
+        self._kmeans = _KMeans(k, metric_type, max_iters=max_iters)
+
+    def __repr__(self) -> str:
+        return f"lance.KMeans(k={self.k}, metric_type={self._metric_type})"
+
+    @property
+    def centroids(self) -> Optional[pa.FixedSizeListArray]:
+        """Returns the centroids of the model."""
+        ret = self._kmeans.centroids()
+        return ret
+
+    def _to_fixed_size_list(self, data: pa.Array) -> pa.FixedSizeListArray:
+        if isinstance(data, pa.FixedSizeListArray):
+            if data.value_type != pa.float32():
+                raise ValueError(f"Array must be float32 type, got: {data.value_type}")
+            return data
+        elif isinstance(data, pa.FixedShapeTensorArray):
+            return pa.FixedSizeListArray.from_arrays(data.storage(), data.shape[1])
```

Comment:
Should we validate the shape of the tensor column here too? Also validate the value type is float32?

Reply:
Done